# Hotel Booking EDA

## Objective
The goal of this exploratory data analysis (EDA) is to understand booking behavior,
identify key factors affecting cancellations, assess data quality,
and extract insights that can guide feature engineering and predictive modeling.


## Importing the libraries

In [2]:
import pandas as pd
import numpy as np
import plotly.express as px

## Load the data

In [3]:
df = pd.read_csv(r"../data/raw/Hotel Reservations.csv")
df.head()

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0   INN00001             2               0                     1   
1   INN00002             2               0                     2   
2   INN00003             1               0                     2   
3   INN00004             2               0                     0   
4   INN00005             2               0                     1   

   no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                  2       Meal Plan 1                           0   
1                  3      Not Selected                           0   
2                  1       Meal Plan 1                           0   
3                  2       Meal Plan 1                           0   
4                  1      Not Selected                           0   

  room_type_reserved  lead_time  arrival_year  arrival_month  arrival_date  \
0        Room_Type 1        224          2017             10             2   
1        Room_Type 1          5          2018             11             6   
2        Room_Type 1          1          2018              2            28   
3        Room_Type 1        211          2018              5            20   
4        Room_Type 1         48          2018              4            11   

  market_segment_type  repeated_guest  no_of_previous_cancellations  \
0             Offline               0                             0   
1              Online               0                             0   
2              Online               0                             0   
3              Online               0                             0   
4              Online               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                       0       Canceled  
3                       0       Canceled  
4                       0       Canceled

The dataset contains hotel reservation records with customer, booking,
and historical information.



## Understanding the data

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            36275 non-null  str    
 1   no_of_adults                          36275 non-null  int64  
 2   no_of_children                        36275 non-null  int64  
 3   no_of_weekend_nights                  36275 non-null  int64  
 4   no_of_week_nights                     36275 non-null  int64  
 5   type_of_meal_plan                     36275 non-null  str    
 6   required_car_parking_space            36275 non-null  int64  
 7   room_type_reserved                    36275 non-null  str    
 8   lead_time                             36275 non-null  int64  
 9   arrival_year                          36275 non-null  int64  
 10  arrival_month                         36275 non-null  int64  
 11  arrival_date              

### Initial Observations
- The dataset contains both numerical and categorical features.
- Some features represent booking behavior (lead_time, special_requests),
  while others describe customer history.
- The target variable is `booking_status`, which indicates whether a booking was canceled.


## Column Overview

The dataset consists of customer-related, booking-related, and historical features.
A brief description of each column is provided to clarify its role in the analysis.
## Feature Description

| Feature                              | Description |
|-------------------------------------|-------------|
| Booking_ID                          | Unique booking identifier (not useful for modeling) |
| no_of_adults                        | Number of adults in the reservation |
| no_of_children                      | Number of children in the reservation |
| no_of_weekend_nights                | Number of weekend nights booked |
| no_of_week_nights                   | Number of weekday nights booked |
| type_of_meal_plan                   | Selected meal plan |
| required_car_parking_space          | Whether a car parking space is required |
| room_type_reserved                  | Type of room reserved |
| lead_time                           | Number of days between booking and arrival |
| arrival_year / arrival_month / arrival_date | Arrival date information |
| market_segment_type                 | Booking market segment |
| repeated_guest                      | Indicates if the guest is a returning customer |
| no_of_previous_cancellations        | Number of previous canceled bookings |
| no_of_previous_bookings_not_canceled| Number of previous successful (non-canceled) bookings |
| avg_price_per_room                  | Average price per room per night |
| no_of_special_requests              | Number of special requests made by the guest |
| booking_status                      | Target variable (Canceled / Not_Canceled) |


In [5]:
# summary numeric statistics
df.describe(include='number').round(2)

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
count      36275.00        36275.00              36275.00           36275.00   
mean           1.84            0.11                  0.81               2.20   
std            0.52            0.40                  0.87               1.41   
min            0.00            0.00                  0.00               0.00   
25%            2.00            0.00                  0.00               1.00   
50%            2.00            0.00                  1.00               2.00   
75%            2.00            0.00                  2.00               3.00   
max            4.00           10.00                  7.00              17.00   

       required_car_parking_space  lead_time  arrival_year  arrival_month  \
count                    36275.00   36275.00      36275.00       36275.00   
mean                         0.03      85.23       2017.82           7.42   
std                          0.17      85.93          0.38           3.07   
min                          0.00       0.00       2017.00           1.00   
25%                          0.00      17.00       2018.00           5.00   
50%                          0.00      57.00       2018.00           8.00   
75%                          0.00     126.00       2018.00          10.00   
max                          1.00     443.00       2018.00          12.00   

       arrival_date  repeated_guest  no_of_previous_cancellations  \
count      36275.00        36275.00                      36275.00   
mean          15.60            0.03                          0.02   
std            8.74            0.16                          0.37   
min            1.00            0.00                          0.00   
25%            8.00            0.00                          0.00   
50%           16.00            0.00                          0.00   
75%           23.00            0.00                          0.00   
max           31.00            1.00                         13.00   

       no_of_previous_bookings_not_canceled  avg_price_per_room  \
count                              36275.00            36275.00   
mean                                   0.15              103.42   
std                                    1.75               35.09   
min                                    0.00                0.00   
25%                                    0.00               80.30   
50%                                    0.00               99.45   
75%                                    0.00              120.00   
max                                   58.00              540.00   

       no_of_special_requests  
count                36275.00  
mean                     0.62  
std                      0.79  
min                      0.00  
25%                      0.00  
50%                      0.00  
75%                      1.00  
max                      5.00

### Key Statistical Observations

- Most bookings are made for **2 adults**, with a very low average number of children,
  indicating that the hotel mainly serves couples or small families.

- The number of **weekend nights** is generally lower than weekday nights,
  suggesting that stays are more common during the working week.

- The `lead_time` variable shows a **high variance** (std ≈ 86 days) with values ranging
  from same-day bookings to reservations made over a year in advance,
  indicating diverse booking behaviors.

- The vast majority of guests **do not require car parking space**, which may reflect
  either the hotel's location or guest travel preferences.

- Most bookings are concentrated in **2017–2018**, with arrival months skewed toward
  mid to late year, suggesting potential seasonality.

- The `repeated_guest` feature is highly imbalanced, with most guests being first-time visitors.

- Historical booking behavior (`no_of_previous_cancellations` and
  `no_of_previous_bookings_not_canceled`) is zero for most customers,
  while a small number of guests exhibit extreme values,
  indicating potential outliers or loyal customer segments.

- The `avg_price_per_room` distribution is **right-skewed**, with most prices clustered
  between 80 and 120, but with extreme high-price outliers reaching up to 540.

- The number of special requests is generally low, with most bookings having
  zero or one request.



In [6]:
# summary categorical statistics
df.describe(include='O')

C:\Users\work\AppData\Local\Temp\ipykernel_39900\3279026830.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='O')


Booking_ID type_of_meal_plan room_type_reserved market_segment_type  \
count       36275             36275              36275               36275   
unique      36275                 4                  7                   5   
top      INN00001       Meal Plan 1        Room_Type 1              Online   
freq            1             27835              28130               23214   

       booking_status  
count           36275  
unique              2  
top      Not_Canceled  
freq            24390

### Categorical Features – Key Observations

- `Booking_ID` contains unique values for each record, confirming that it serves
  purely as an identifier and does not provide predictive value.

- The `type_of_meal_plan` feature has a dominant category, with *Meal Plan 1*
  accounting for the majority of bookings, indicating a strong customer preference
  or a default option during booking.

- `room_type_reserved` is highly skewed toward *Room_Type 1*, suggesting that
  standard room types are the most frequently booked.

- The `market_segment_type` feature is dominated by the *Online* segment,
  highlighting the importance of online booking channels.

- The target variable `booking_status` shows that most bookings are not canceled,
  indicating a moderate class imbalance.


# Data Quality Checks

In [7]:
(df.isnull().mean() * 100).round(2)

Booking_ID                              0.0
no_of_adults                            0.0
no_of_children                          0.0
no_of_weekend_nights                    0.0
no_of_week_nights                       0.0
type_of_meal_plan                       0.0
required_car_parking_space              0.0
room_type_reserved                      0.0
lead_time                               0.0
arrival_year                            0.0
arrival_month                           0.0
arrival_date                            0.0
market_segment_type                     0.0
repeated_guest                          0.0
no_of_previous_cancellations            0.0
no_of_previous_bookings_not_canceled    0.0
avg_price_per_room                      0.0
no_of_special_requests                  0.0
booking_status                          0.0
dtype: float64

### Missing Values Analysis

The dataset does not contain missing values across any of the features.
This indicates good data quality and removes the need for imputation strategies.


In [8]:
df.duplicated().sum()

np.int64(0)

### Duplicate Records Check

No duplicated rows were found in the dataset.
Each booking record appears to be unique.


## Univariate Analysis (numerical)

In [9]:
num_cols = df.select_dtypes(include=["int64", "float64"]).columns.drop(['repeated_guest', 'required_car_parking_space'])
num_cols

Index(['no_of_adults', 'no_of_children', 'no_of_weekend_nights',
       'no_of_week_nights', 'lead_time', 'arrival_year', 'arrival_month',
       'arrival_date', 'no_of_previous_cancellations',
       'no_of_previous_bookings_not_canceled', 'avg_price_per_room',
       'no_of_special_requests'],
      dtype='str')

In [10]:
for col in num_cols:
    px.histogram(df, x=col).show()

## Univariate Analysis (categorical)

In [11]:
df.columns

Index(['Booking_ID', 'no_of_adults', 'no_of_children', 'no_of_weekend_nights',
       'no_of_week_nights', 'type_of_meal_plan', 'required_car_parking_space',
       'room_type_reserved', 'lead_time', 'arrival_year', 'arrival_month',
       'arrival_date', 'market_segment_type', 'repeated_guest',
       'no_of_previous_cancellations', 'no_of_previous_bookings_not_canceled',
       'avg_price_per_room', 'no_of_special_requests', 'booking_status'],
      dtype='str')

In [12]:
cat_cols = ['required_car_parking_space','type_of_meal_plan','room_type_reserved','market_segment_type','repeated_guest','booking_status']
cat_cols

['required_car_parking_space',
 'type_of_meal_plan',
 'room_type_reserved',
 'market_segment_type',
 'repeated_guest',
 'booking_status']

In [13]:
for col in cat_cols:
    counts = df[col].value_counts().reset_index()
    counts.columns = [col, 'count']
    px.bar(counts, x=col, y='count', title=f'Value counts of {col}').show()

## Bivariate Analysis with Target Variable

In [14]:
df.columns

Index(['Booking_ID', 'no_of_adults', 'no_of_children', 'no_of_weekend_nights',
       'no_of_week_nights', 'type_of_meal_plan', 'required_car_parking_space',
       'room_type_reserved', 'lead_time', 'arrival_year', 'arrival_month',
       'arrival_date', 'market_segment_type', 'repeated_guest',
       'no_of_previous_cancellations', 'no_of_previous_bookings_not_canceled',
       'avg_price_per_room', 'no_of_special_requests', 'booking_status'],
      dtype='str')

### Lead Time vs Booking Status

In [15]:
fig = px.box(df, x='booking_status', y='lead_time', color='booking_status', title='Lead Time by Booking Status')
fig.show()


* **Lead Time Impact :** Canceled bookings show a significantly higher median lead time compared to confirmed ones, indicating that the further in advance a room is booked, the higher the risk of cancellation.
* **Risk Mitigation :** Bookings exceeding 150 days of lead time are high-risk; implementing stricter deposit policies for these cases could help reduce the cancellation rate.



### Market Segment vs Booking Status

In [16]:
fig = px.histogram(df, x='market_segment_type', color='booking_status', 
                   barmode='group', title='Market Segment Type by Booking Status')
fig.show()


* **Segment Trends:** Online bookings represent the highest volume of both total and canceled reservations, suggesting a need for stricter deposit policies for long-term online bookings.



### Repeated Guest vs Booking Status

In [17]:
px.bar(
    df.groupby(["repeated_guest", "booking_status"])
      .size()
      .reset_index(name="count"),
    x="repeated_guest",
    y="count",
    color="booking_status",
    barmode="group",
    title="Repeated Guest vs Booking Status"
).show()



* **New Guest Volatility:** First-time guests (value 0) have significantly higher cancellation rates compared to repeat guests.
* **High Retention in Loyal Guests:** Repeat guests (value 1) show a nearly zero cancellation rate; almost all their bookings remain confirmed.



In [27]:
df_avg = df.groupby('room_type_reserved')['avg_price_per_room'].mean().reset_index()

fig = px.bar(df_avg, x='room_type_reserved', y='avg_price_per_room',
             text_auto='.2f', 
             title='Average Price per Room by Room Type Reserved')
fig.show()



* **Average Room Price Across Different Room Types**
Room_Type 6 and Room_Type 7 have the highest average prices, indicating premium room offerings, while Room_Type 3 is the most budget-friendly. This clear price segmentation across room types reflects differences in amenities and can influence both booking behavior and cancellation patterns.


In [19]:
px.bar(
    df.groupby(["room_type_reserved", "booking_status"])
      .size()
      .reset_index(name="count"),
    x="room_type_reserved",
    y="count",
    color="booking_status",
    barmode="group",
    title="Room Type vs Booking Status"
).show()

In [29]:
df_mean = df.groupby("booking_status")["avg_price_per_room"].mean().reset_index()

fig = px.bar(
    df_mean, 
    x="booking_status", 
    y="avg_price_per_room", 
    color="booking_status",
    text_auto='.2f', 
    title="Average Price per Room by Booking Status"
)

fig.show()

* **Higher room prices** are directly correlated with higher cancellation rates, showing a **10.7% price premium** on canceled bookings ($110.59 vs ).
  This suggests guests are more prone to "buyer’s remorse" or price-shopping at higher tiers, making **non-refundable rates** a vital strategy for high-value bookings.



In [23]:
monthly_price = df.groupby('arrival_month')['avg_price_per_room'].mean().reset_index()
fig = px.line(monthly_price, x='arrival_month', y='avg_price_per_room', 
              markers=True, title='Average Price per Room by Arrival Month')
fig.show()



* **Seasonal Price Sensitivity**
Room prices peak between May and September, and canceled bookings tend to have a higher average price than fulfilled ones. This suggests that high seasonal pricing is associated with an increased risk of cancellations.




In [ ]:
fig = px.scatter(df, x='lead_time', y='avg_price_per_room',
                 color='booking_status', opacity=0.5,
                 title='Relationship between Lead Time and Room Price')
fig.show()



* **Lead Time, Price, and Cancellation Behavior**
Room prices decrease as lead time increases, while cancellations are more common for early bookings and less frequent for high-priced, last-minute reservations.


.


## **Comprehensive Hotel Booking Analysis Report**

### **1. Numerical Features Analysis**

**Timing & Stay Patterns (`lead_time`, `week_nights`, `weekend_nights`)**

* **Short Stays:** Most guests book 1–3 weeknights and reserve less than 50 days in advance.
* **Lead Time Impact:** Canceled bookings have a significantly higher median lead time compared to confirmed ones, indicating that bookings made further in advance carry a higher risk of cancellation.
* **Risk Mitigation:** Bookings exceeding 150 days of lead time are high-risk; implementing stricter deposit policies for these cases could help reduce cancellations.

**Guest Profile & Loyalty (`adults`, `children`, `previous_cancellations`, `repeated_guest`)**

* **Primary Segment:** Couples (2 adults, 0 children) dominate bookings.
* **New Guest Volatility:** First-time guests have significantly higher cancellation rates compared to repeat guests.
* **High Retention in Loyal Guests:** Repeat guests show a nearly zero cancellation rate; almost all of their bookings remain confirmed.
* **Customer Trust:** Guests with previous cancellations are higher-risk clients.

**Revenue & Engagement (`avg_price_per_room`, `special_requests`, `room_type_reserved`)**

* **Pricing Trends:**

  * Room_Type 6 and Room_Type 7 have the highest average prices (premium rooms), while Room_Type 3 is the most budget-friendly.
  * Higher room prices are correlated with higher cancellation rates—a **10.7% premium** on canceled bookings ($110.59 vs confirmed).
  * Recommendation: Implement **non-refundable rates** for high-value bookings to reduce “buyer’s remorse.”
* **Seasonal Price Sensitivity:** Room prices peak from May to September; canceled bookings tend to have higher prices during these months.
* **Lead Time & Price Behavior:** Room prices decrease as lead time increases. Cancellations are more common for early bookings and less frequent for high-priced, last-minute reservations.
* **Engagement:** Guests with more special requests tend to cancel less, indicating higher commitment.

**Arrival Trends (`arrival_year`, `month`, `date`)**

* **Growth:** Booking volume surged in 2018 compared to 2017.
* **Seasonality:** October experiences peak arrivals; January sees a dip.



### **2. Categorical Features Analysis**

**Booking Channels & Status (`market_segment_type`, `booking_status`)**

* **Dominant Channel:** Online bookings are the main source of traffic.
* **Segment Trends:** Online bookings have the highest volume of both total and canceled reservations.
* **Cancellation Risk:** About **33%** of bookings are canceled, suggesting focus on retention strategies and stricter policies for long-term online bookings.

**Preferences & Services (`type_of_meal_plan`, `required_car_parking_space`)**

* **Meal Choice:** **Meal Plan 1** is the most popular.
* **Low Parking Demand:** Less than 5% of guests require parking, implying reliance on public transport or ride-sharing.

**Room Selection & Loyalty (`room_type_reserved`, `repeated_guest`)**

* **Popular Room Type:** **Room_Type 1** is most reserved, indicating strong preference for standard rooms.
* **Guest Retention:** The majority are new guests; repeat bookings are very low, highlighting a gap in loyalty.



### **3. Key Insights & Recommendations**

1. **High-Risk Bookings:**

   * Focus on early bookings with long lead times (>150 days).
   * Consider stricter deposit policies or non-refundable rates for premium/high-priced rooms.

2. **Retention Opportunities:**

   * Target first-time guests with loyalty programs to increase repeat bookings.
   * Online booking channels require careful monitoring due to higher cancellation rates.

3. **Pricing & Engagement:**

   * Maintain data quality for room prices.
   * Use special requests as a proxy for guest commitment.
   * Monitor seasonal pricing carefully to avoid price-sensitive cancellations.

4. **Operational Planning:**

   * Prepare for peak season (October) with staffing and resource allocation.
   * Adjust services (meal plans, parking) based on actual demand.

5. **Revenue Optimization:**

   * Price segmentation across room types allows targeted marketing strategies.
   * Premium room cancellations suggest the need for non-refundable or deposit-based bookings to secure revenue.



